### PR TITLE
Myshop 5

### PR DESCRIPTION
**Pull request by nvtrinh2412**
- sort, paginate, find product
- test for productService

### CODE REVIEW

File: backend/file/data/myshop.trace.db [hunk 1]
```diff
@@ -0,0 +1,171 @@
+2022-08-08 07:26:00 jdbc[3]: exception
```
Comment (thangthnguyen-kms) [inline backend/file/data/myshop.trace.db]: Do we need this file for the app to run properly? if not then ignore this for the next PR (pull request)

File: backend/src/main/java/com/myshop/products/ProductController.java [hunk 2]
```diff
@@ -21,46 +22,80 @@ public class ProductController {
...
+            List<Product> products = productService.getAllProduct();
+            return new ResponseEntity<>(new SuccessfulResponse<>(products.size(), products), HttpStatus.OK);
+
+        } catch (Exception e) {
+            return new ResponseEntity<>(new FailureResponse("Internal Server Error"), HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    @GetMapping(params = {"name"})
+    public ResponseEntity<ResponseData> getAllProductOfName(@RequestParam String name) {
```
Comment (thangthnguyen-kms) [inline backend/src/main/java/com/myshop/products/ProductController.java]: i think the correct term should be `getAllProductByName`

File: backend/src/main/java/com/myshop/products/ProductController.java [hunk 3]
```diff
@@ -21,46 +22,80 @@ public class ProductController {
...
+            return new ResponseEntity<>(new SuccessfulResponse<>(products.size(), products), HttpStatus.OK);
+
+        } catch (Exception e) {
+            return new ResponseEntity<>(new FailureResponse("Internal Server Error"), HttpStatus.INTERNAL_SERVER_ERROR);
         }
+    }
 
-        // get all product
+    @GetMapping(params = {"page","size"})
+    public ResponseEntity<ResponseData> getAllProductWithPaginate(@RequestParam String page, @RequestParam String size) {
```
Comment (thangthnguyen-kms) [inline backend/src/main/java/com/myshop/products/ProductController.java]: i don't think this is the correct Type for page and size parameter.
 what will happen if i use: **Page** = '**a12**' and **size** = '**2b3**'

File: backend/src/test/java/com/myshop/products/ProductServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        // 4. verify result
+        assertThat(result.size()).isEqualTo(productsList.size());
+        // 5. verify behavior of mock
+        verify(productRepository).findProductByName(any(String.class));
+    }
+
+    @Test
+    void paginate() {
+    }
+}
```
Comment (thangthnguyen-kms) [inline backend/src/test/java/com/myshop/products/ProductServiceTest.java:147]: git remind you to add a new line at the end of file

File: backend/src/test/java/com/myshop/products/ProductServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+    @Test
+    void addNewProduct() {
+        Product product = new Product("Testing", 100, "image", List.of("red", "blue", "green"), "This is a test product");
+        Product addedProduct = productRepository.save(product);
+
+        ArgumentCaptor<Product> argument = ArgumentCaptor.forClass(Product.class);
+        verify(productRepository).save(argument.capture());
+        Product capturedProduct = argument.getValue();
+        System.out.println(productRepository.findAll());
```
Comment (thangthnguyen-kms) [inline backend/src/test/java/com/myshop/products/ProductServiceTest.java]: remove unnecessary code before create PR. 
Ex: comment, sop, ...

**Review by thangthnguyen-kms — COMMENTED** [review]
Great work with the unit testing :D

File: backend/src/main/java/com/myshop/products/ProductController.java (hunk 3 above)
Comment (nvtrinh2412) [reply backend/src/main/java/com/myshop/products/ProductController.java]: In productService, I tried to parse these parameters to int. So with the case **Page = 'a12' and size = '2b3'**,  DEFAULT_VALUE will be applied to them. 
If I use type **int** for aforementioned params, "Bad Request" error will occur with your case

**Comment by haibuiminh** [comment]
1. file log issues like 
```
backend/file/data/myshop.trace.db
``` 
should not push to github

**Comment by haibuiminh** [comment]
2. Code convention comment:
Screen Shot 2022-08-15 at 10 09 05 AM
you should install the sonar lint plugin into IntelliJ IDEA.
https://www.sonarsource.com/products/sonarlint/?gads_campaign=North-America-SonarSource&gads_ad_group=SonarSource&gads_keyword=sonarsource&gclid=Cj0KCQjwuuKXBhCRARIsAC-gM0hJLOmW1FbakAj_GuIPHz5fN3CW3dstcbdkAwNy0zcRvEwZoZZxQ6YaAlsREALw_wcB
Then install dependencies spotless for this project to auto-fix convention issues
https://github.com/diffplug/spotless/tree/main/plugin-maven

**Comment by haibuiminh** [comment]
3. database design
```
@Column(name = "name_id", nullable = false, unique = true)
    private String nameId;
```
why we need nameId? we can use id instead

**Comment by haibuiminh** [comment]
4. add index to entity to improve performance of search 
https://www.baeldung.com/jpa-indexes

**Comment by haibuiminh** [comment]
5. Controller with request param
using validate
https://www.baeldung.com/spring-validate-requestparam-pathvariable

**Comment by haibuiminh** [comment]
6. Service pagination 
https://www.baeldung.com/spring-data-jpa-pagination-sorting

**Comment by haibuiminh** [comment]
7. default page size 20

**Comment by haibuiminh** [comment]
8. 
Screen Shot 2022-08-15 at 10 24 38 AM
u need to implement the controller simple as this. let the complex part of the service
Screen Shot 2022-08-15 at 10 25 36 AM
Screen Shot 2022-08-15 at 10 27 24 AM

**Comment by haibuiminh** [comment]
9. 
```
ResponseEntity<ResponseData>
``` 
we do not need to use it. let the java spring handle HTTP code status

for example 
```
 @GetMapping(params = {"page","size"})
    public Page<Product> getAllProductWithPaginate(@RequestParam @Min(0) Integer page, @RequestParam @Min(20) Integer size) {
```

**Comment by haibuiminh** [comment]
10. sort and pagination in only function 
```
Page size and Page id always NUMBER, Integer, Long
```
https://www.baeldung.com/spring-data-jpa-pagination-sorting#sort

**Comment by haibuiminh** [comment]
11. just using 
```
Optional<Product> findProductByNameId(String nameId);
```
instead 
Screen Shot 2022-08-15 at 10 35 22 AM

**Comment by haibuiminh** [comment]
12. Just using lombok when you know about "lombok what did it do behind the seen?"

**Comment by haibuiminh** [comment]
12. for example. u should write code like this with @Autowired for constructor 
Screen Shot 2022-08-15 at 10 41 46 AM

https://www.baeldung.com/constructor-injection-in-spring

**Comment by haibuiminh** [comment]
13. Remove redundant code

**Comment by haibuiminh** [comment]
14. Add Document for api using open api https://www.baeldung.com/spring-rest-openapi-documentation

**Comment by nvtrinh2412** [comment]
> 3. database design
> 
> ```
> @Column(name = "name_id", nullable = false, unique = true)
>     private String nameId;
> ```
> 
> why we need nameId? we can use id instead

Because when I see on the demo app (https://demo.vercel.store/search). I saw that the slug in the URL look like name of the product. 
image
Therefore I think if my shop had 2 products with the same name, they would have a same slug in URL. So I use nameId to avoid that case.

**Pull request merged**